synapse/visibility.py

```diff
@@ -645,6 +646,17 @@ def check_event_is_visible(
         visible = check_event_is_visible(
             event_to_history_vis[e.event_id], event_to_memberships.get(e.event_id, {})
         )
+
+        # Filter out non-local events when we are in the middle of a partial join,
+        # since our servers list can be out of date and we could leak events
+        # to servers not in the room anymore.
+        # This can also be true for local events but we consider it to be
+        # an acceptable risk in this case.
+        if e.origin != local_server_name and await storage.main.is_partial_state_room(
+            e.room_id
+        ):
```

Comment:
There's a minor race condition here where we get a list of members for a partial state room which contains an incorrectly accepted membership, then by the time we go to check the partial state flag, the room has been un partial stated and we expose the event based on that incorrect membership.

I think we need to collect all the partial state flags together first, before the call to `_event_to_memberships`.